File: nim/solver.nim
import algorithm, strformat, strutils, sugar, tables, timeit

const letters = "abcdefg"

proc isValid(word: string): bool =
  for letter in word:
    if word.count(letter) > letters.count(letter):
      return false
    
  true

proc loadValidWords(): seq[string] =
  for line in readFile("dictionary.txt").split("\r\n"):
    if line.isValid():
      result.add line

proc getScore(word: string): int =
  const scores = [1, 3, 3, 2, 1, 4, 2, 4, 1, 8, 5, 1, 3, 1, 1, 3, 10, 1, 
    1, 1, 11, 4, 4, 8, 4, 10]

  for letter in word:
    result += scores[letter.byte - 97]

func sortWords(words: var seq[string]) =
  words.sort((a, b) => (
    if a.getScore() < b.getScore(): -1 else: 1
  ))

proc displayWords(words: seq[string]) =
  for i in countdown(words.len, 1):
    echo &"{i}. {words[^i]}"

echo timeGo do:
  var words = loadValidWords()
  sortWords words
  displayWords words

File: julia/solver.jl
using Printf

function is_valid(word)
    for letter in word
        if count(c -> (c == letter), word) > count(c -> (c == letter), letters)
            return false
        end
    end

    return true
end

function load_valid_words()
    return [line for line in readlines("dictionary.txt") if is_valid(line)]
end

const global scores = Dict(
	'a' => 1, 'b' => 3, 'c' => 3, 'd' => 2,
	'e' => 1, 'f' => 4, 'g' => 2, 'h' => 4,
	'i' => 1, 'j' => 8, 'k' => 5, 'l' => 1,
	'm' => 3, 'n' => 1, 'o' => 1, 'p' => 3,
	'q' => 10,'r' => 1, 's' => 1, 't' => 1,
	'u' => 1, 'v' => 4, 'w' => 4, 'x' => 8,
    'y' => 4, 'z' => 10
)

function get_score(word)
    return sum(scores[letter] for letter in word)
end

function display_words(words) 
    len = length(words)

    for i = len:-1:1
        index = len - i + 1
        @printf("%i. %s\n", i, words[index])
    end
end

const global letters = "abcdefg"

function main()
    words = load_valid_words()
    sort!(words, by=get_score)
    display_words(words)
end
